numpy/ma/core.py

```diff
@@ -3184,13 +3180,15 @@ def __setitem__(self, indx, value):
 
         """
         # 2016.01.15 -- v1.11.0
-        warnings.warn(
-           "Currently, slicing will try to return a view of the data," +
-           " but will return a copy of the mask. In the future, it will try" +
-           " to return both as views. This means that using `__setitem__`" +
-           " will propagate values back through all masks that are present.",
-           FutureWarning
-        )
+        if (self._mask is not nomask) and \
+                (self._sharedmask or not self._data.flags.owndata):
```

Comment:
Right, as views on the data act only as a proxy for what we really want to know. I agree reference counting is to messy.

One thought I had is if we have some flag set here in `__setitem__` to note if the mask was shared previously like `self._oldsharedmask`. Admittedly, if other references to the old mask are dropped, there might still be a few false positives, but trying to address that special case involves referenced counting, which I think we both want to avoid.